src/ray/raylet/node_manager.cc

```diff
@@ -736,6 +741,73 @@ void NodeManager::DispatchTasks(
   local_queues_.MoveTasks(assigned_task_ids, TaskState::READY, TaskState::RUNNING);
 }
 
+std::pair<std::shared_ptr<Worker>, bool> NodeManager::GetWorker(
+    const WorkerID &worker_id) {
+  bool is_worker = true;
+  auto worker = worker_pool_.GetRegisteredWorker(worker_id);
+  if (!worker) {
+    is_worker = false;
+    worker = worker_pool_.GetRegisteredDriver(worker_id);
+    if (!worker) {
+      return std::make_pair(nullptr, true);
+    }
+  }
+  return std::make_pair(worker, is_worker);
+}
+
+#define PREPROCESS_REQUEST(REQUEST_TYPE)                                            \
+  WorkerID worker_id = WorkerID::FromBinary(request.worker_id());                   \
+  do {                                                                              \
+    if (!PreprocessRequest<rpc::REQUEST_TYPE>(worker_id, #REQUEST_TYPE, request)) { \
+      return;                                                                       \
+    }                                                                               \
+  } while (0)
+
+template <typename Request>
+bool NodeManager::PreprocessRequest(const WorkerID &worker_id,
+                                    const std::string &request_name,
+                                    const Request &request) {
+  std::ostringstream string_stream;
+  string_stream << "Received a " << request_name << " request. Worker id "
+                << worker_id.Hex() << ".";
+  // NOTE(Joey Jiang): We check heartbeat actively because the heartbeats' callback won't
+  // be invoked until all the post requests in the grpc server have been finished.
+  // The raylet might be marked as dead due to missing heartbeats when too many requests
+  // are sent to this raylet.
+  int64_t expiry_delay =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::steady_clock::now() - heartbeat_timer_.expires_at())
+          .count();
+  int64_t expiry_delay_limit = RayConfig::instance().heartbeat_timeout_milliseconds() *
+                               RayConfig::instance().num_heartbeats_timeout() / 2;
+  // Timer's callback hasn't been called on the desired time.
+  if (expiry_delay > expiry_delay_limit) {
```

Comment:
do we need to check this for every request? Maybe reduce it to every 10 requests to reduce the time calculation cost?

Reply:
I think it should be very cheap to calculate this?